passes/hierarchy/hierarchy.cc

```diff
@@ -376,6 +376,65 @@ RTLIL::Module *get_module(RTLIL::Design                  &design,
 	return nullptr;
 }
 
+// Try to read an IdString as a numbered connection name ("$123" or similar),
+// writing the result to dst. If the string isn't of the right format, ignore
+// dst and return false.
+bool read_id_num(RTLIL::IdString str, int *dst)
+{
+	log_assert(dst);
+
+	const char *c_str = str.c_str();
+	if (c_str[0] != '$' || !('0' <= c_str[1] && c_str[1] <= '9'))
+		return false;
+
+	*dst = atoi(c_str + 1);
+	return true;
+}
+
+// Check that all the connections and parameters on the cell match those defined
+// in its type.
```

Comment:
I think I'd like this comment to clarify that it's checking the connection/parameter _names and positions_ to make sure they exist, as "match" is a little ambiguous at first glance.